### PR TITLE
Ps/pm 5533/migrate decrypted user key

### PR DESCRIPTION
**Pull request by MGibson1**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

- Migrates in-memory user key to state providers.
- Creates new crypto service for browser rather than taking if from the background 🥳 
- Adds the concept of a downgrade-only update to authentication status. This is important because we don't want to have to worry about the ordering of calls that cause lock vs logout.
    - ❓ **Should we change lock/unlock to be a boolean `isUnlocked` and logged out/ logged in to a `hasToken` boolean? The observable could then calculate the authentication status if we feel that's still a useful tool.** @jlf0dev your input would be valuable here.

## Code changes

- **browser services module**: Creates new crypto service and crypto function service specific to the foreground.  The crypto service's factory is to register crypto service and encrypt service to the bitContainerService for decryption purposes (working on refactoring this out, but it's there for now).
- **crypto service**: 
  - Added observable for the active user key.
  - Updated set user key to explicitly throw on nullish keys (it already would if you were storing an auto key, but now it has a better error message and documentation)
  - Set account auth status based on presence of user keys
- **state service**: Deprecated the active account unlocked observable and tied to the account service as the new source of truth. Crypto service is now directly updating account statuses based on presence of user keys.

## Test Stuff
The jest config update just adds the common spec directory to jest's paths without messing with tsconfig paths, since it's a test-only concern.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

**Review by jlf0dev — COMMENTED** [review]
> ❓ Should we change lock/unlock to be a boolean isUnlocked and logged out/ logged in to a hasToken boolean? The observable could then calculate the authentication status if we feel that's still a useful tool.

If I'm following this correctly, you're talking about having authentication status be derived from two pieces of state? Would the `hasToken` boolean be from the `TokenService`? Because that inverts the dependency I think we were heading with state being dependent on the authentication status. Is this question coming from our struggles with subscriptions in services?

**Comment by MGibson1** [comment]
@jlf0dev yes and yes. Token service would be the source of truth of whether or not you're authed (because that is the definition of it, right?) and this would be the source of truth of whether or not your unlocked.

I wasn't necessarily thinking about your token service dependency issues, but more that as we migrate to being more reactive, we can't follow a clear path on the road from logged out to unlocked and one way to solve that would be to separate the state. The way implemented with a max status here works too, but is a little odd in an interface IMO

**Comment by jlf0dev** [comment]
@MGibson1 Thats fair, I think deriving the authentication status does prove that it will never be out of sync. I think this is a good idea as long as we try to route everything through the account service still. It should be the interface for the rest of the application.

This means we should probably look into limiting the use of the token service moving forward. Ideally it should be Auth internal and not used in components.

CC: @JaredSnider-Bitwarden since you've been working directly in this area with #7975.

File: libs/common/src/platform/services/state.service.ts [hunk 1]
```diff
@@ -109,22 +109,11 @@ export class StateService<
...
-          //  implementation detail.
-          this.activeAccountUnlockedSubject.next((await this.getUserKey()) != null);
-        }),
-      )
-      .subscribe();
+    this.activeAccountUnlocked$ = this.accountService.activeAccount$.pipe(
+      map((a) => {
+        return a?.status === AuthenticationStatus.Unlocked;
+      }),
+    );
```
[the remark this answers is not in the document]
Comment (MGibson1) [reply libs/common/src/platform/services/state.service.ts:116]: Nope, that is used in the decryption of user keys with master keys, which is the domain of a different service, your master key service iirc.

File: libs/common/src/platform/services/state.service.ts [hunk 2]
```diff
@@ -109,22 +108,11 @@ export class StateService<
...
-          } else if (userId == null) {
-            this.activeAccountUnlockedSubject.next(false);
-          }
-          // FIXME: This should be refactored into AuthService or a similar service,
-          //  as checking for the existence of the crypto key is a low level
-          //  implementation detail.
-          this.activeAccountUnlockedSubject.next((await this.getUserKey()) != null);
-        }),
-      )
-      .subscribe();
```
Comment (justindbaur) [inline libs/common/src/platform/services/state.service.ts:127]: Early subscribe gone ☺️

**Review by justindbaur — APPROVED** [review]
Looks great.

**Review by jlf0dev — APPROVED** [review]

**Pull request merged**